### PR TITLE
fix raising NotYetImplementedError against commands which exist in image-element

### PR DESCRIPTION
**Pull request by KazuCocoa**
When I confirmed _add -image locator strategy and related logic #235_ in https://github.com/appium/ruby_lib_core/pull/119/files , I faced below error.

```
[debug] [W3C] Calling AppiumDriver.getElementRect() with args: ["appium-image-element-c936dd11-9589-40fe-8a63-5c4ac2cea22f","d637cc26-2895-4e3a-976f-40c8baf5f425"]
[debug] [XCUITest] Executing command 'getElementRect'
[W3C] Encountered internal error running command:  {"jsonwpCode":13,"error":"unknown method","w3cStatus":404,"_stacktrace":null} NotYetImplementedError: Method has not yet been implemented
[W3C]     at XCUITestDriver.executeCommand$ (/Users/kazuaki/GitHub/appium/node_modules/appium-base-driver/lib/basedriver/driver.js:263:13)
[W3C]     at tryCatch (/Users/kazuaki/GitHub/appium/node_modules/babel-runtime/regenerator/runtime.js:67:40)
[W3C]     at GeneratorFunctionPrototype.invoke [as _invoke] (/Users/kazuaki/GitHub/appium/node_modules/babel-runtime/regenerator/runtime.js:315:22)
[W3C]     at GeneratorFunctionPrototype.prototype.(anonymous function) [as next] (/Users/kazuaki/GitHub/appium/node_modules/babel-runtime/regenerator/runtime.js:100:21)
[W3C]     at invoke (/Users/kazuaki/GitHub/appium/node_modules/babel-runtime/regenerator/runtime.js:136:37)
[W3C]     at enqueueResult (/Users/kazuaki/GitHub/appium/node_modules/babel-runtime/regenerator/runtime.js:185:17)
[W3C]     at new Promise (<anonymous>)
[W3C]     at new F (/Users/kazuaki/GitHub/appium/node_modules/core-js/library/modules/$.export.js:30:36)
[W3C]     at AsyncIterator.enqueue (/Users/kazuaki/GitHub/appium/node_modules/babel-runtime/regenerator/runtime.js:184:12)
[W3C]     at AsyncIterator.prototype.(anonymous function) [as next] (/Users/kazuaki/GitHub/appium/node_modules/babel-runtime/regenerator/runtime.js:100:21)
[W3C]     at Object.runtime.async (/Users/kazuaki/GitHub/appium/node_modules/babel-runtime/regenerator/runtime.js:209:12)
[W3C]     at XCUITestDriver.executeCommand (/Users/kazuaki/GitHub/appium/node_modules/appium-base-driver/build/lib/basedriver/driver.js:283:34)
[W3C]     at XCUITestDriver.executeCommand$ (/Users/kazuaki/GitHub/appium/node_modules/appium-xcuitest-driver/lib/driver.js:111:21)
[W3C]     at tryCatch (/Users/kazuaki/GitHub/appium/node_modules/babel-runtime/regenerator/runtime.js:67:40)
[W3C]     at GeneratorFunctionPrototype.invoke [as _invoke] (/Users/kazuaki/GitHub/appium/node_modules/babel-runtime/regenerator/runtime.js:315:22)
[W3C]     at GeneratorFunctionPrototype.prototype.(anonymous function) [as next] (/Users/kazuaki/GitHub/appium/node_modules/babel-runtime/regenerator/runtime.js:100:21)
[W3C]     at invoke (/Users/kazuaki/GitHub/appium/node_modules/babel-runtime/regenerator/runtime.js:136:37)
[W3C]     at enqueueResult (/Users/kazuaki/GitHub/appium/node_modules/babel-runtime/regenerator/runtime.js:185:17)
[W3C]     at new Promise (<anonymous>)
[W3C]     at new F (/Users/kazuaki/GitHub/appium/node_modules/core-js/library/modules/$.export.js:30:36)
[W3C]     at AsyncIterator.enqueue (/Users/kazuaki/GitHub/appium/node_modules/babel-runtime/regenerator/runtime.js:184:12)
[W3C]     at AsyncIterator.prototype.(anonymous function) [as next] (/Users/kazuaki/GitHub/appium/node_modules/babel-runtime/regenerator/runtime.js:100:21)
[W3C]     at Object.runtime.async (/Users/kazuaki/GitHub/appium/node_modules/babel-runtime/regenerator/runtime.js:209:12)
[W3C]     at XCUITestDriver.executeCommand (/Users/kazuaki/GitHub/appium/node_modules/appium-xcuitest-driver/build/lib/driver.js:1204:34)
[W3C]     at AppiumDriver.executeCommand$ (/Users/kazuaki/GitHub/appium/lib/appium.js:425:50)
[W3C]     at tryCatch (/Users/kazuaki/GitHub/appium/node_modules/babel-runtime/regenerator/runtime.js:67:40)
[W3C]     at GeneratorFunctionPrototype.invoke [as _invoke] (/Users/kazuaki/GitHub/appium/node_modules/babel-runtime/regenerator/runtime.js:315:22)
[W3C]     at GeneratorFunctionPrototype.prototype.(anonymous function) [as next] (/Users/kazuaki/GitHub/appium/node_modules/babel-runtime/regenerator/runtime.js:100:21)
[W3C]     at GeneratorFunctionPrototype.invoke (/Users/kazuaki/GitHub/appium/node_modules/babel-runtime/regenerator/runtime.js:136:37)
[W3C]     at <anonymous>
[W3C]     at process._tickCallback (internal/process/next_tick.js:160:7)
[HTTP] <-- GET /wd/hub/session/d637cc26-2895-4e3a-976f-40c8baf5f425/element/appium-image-element-c936dd11-9589-40fe-8a63-5c4ac2cea22f/rect 404 11 ms - 3698
[HTTP]
^C[Appium] Received SIGINT - shutting down
```

After this PR, they worked like below.

```
[HTTP] --> GET /wd/hub/session/fe7279cc-a791-45be-80b6-3308032e909c/element/appium-image-element-bf442481-2c5d-4d36-9033-a358ae6039eb/rect
[HTTP] {}
[debug] [W3C] Calling AppiumDriver.getElementRect() with args: ["appium-image-element-bf442481-2c5d-4d36-9033-a358ae6039eb","fe7279cc-a791-45be-80b6-3308032e909c"]
[debug] [XCUITest] Executing command 'getElementRect'
[debug] [W3C] Responding to client with driver.getElementRect() result: {"x":24,"y":162,"width":170,"height":53}
[HTTP] <-- GET /wd/hub/session/fe7279cc-a791-45be-80b6-3308032e909c/element/appium-image-element-bf442481-2c5d-4d36-9033-a358ae6039eb/rect 200 7 ms - 50
[HTTP] 
[HTTP] --> GET /wd/hub/session/fe7279cc-a791-45be-80b6-3308032e909c/element/appium-image-element-bf442481-2c5d-4d36-9033-a358ae6039eb/rect
[HTTP] {}
[debug] [W3C] Calling AppiumDriver.getElementRect() with args: ["appium-image-element-bf442481-2c5d-4d36-9033-a358ae6039eb","fe7279cc-a791-45be-80b6-3308032e909c"]
[debug] [XCUITest] Executing command 'getElementRect'
[debug] [W3C] Responding to client with driver.getElementRect() result: {"x":24,"y":162,"width":170,"height":53}
[HTTP] <-- GET /wd/hub/session/fe7279cc-a791-45be-80b6-3308032e909c/element/appium-image-element-bf442481-2c5d-4d36-9033-a358ae6039eb/rect 200 6 ms - 50
[HTTP] 
[HTTP] --> GET /wd/hub/session/fe7279cc-a791-45be-80b6-3308032e909c/element/appium-image-element-bf442481-2c5d-4d36-9033-a358ae6039eb/rect
[HTTP] {}
[debug] [W3C] Calling AppiumDriver.getElementRect() with args: ["appium-image-element-bf442481-2c5d-4d36-9033-a358ae6039eb","fe7279cc-a791-45be-80b6-3308032e909c"]
[debug] [XCUITest] Executing command 'getElementRect'
[debug] [W3C] Responding to client with driver.getElementRect() result: {"x":24,"y":162,"width":170,"height":53}
[HTTP] <-- GET /wd/hub/session/fe7279cc-a791-45be-80b6-3308032e909c/element/appium-image-element-bf442481-2c5d-4d36-9033-a358ae6039eb/rect 200 8 ms - 50
[HTTP] 
[HTTP] --> GET /wd/hub/session/fe7279cc-a791-45be-80b6-3308032e909c/element/appium-image-element-bf442481-2c5d-4d36-9033-a358ae6039eb/rect
[HTTP] {}
[debug] [W3C] Calling AppiumDriver.getElementRect() with args: ["appium-image-element-bf442481-2c5d-4d36-9033-a358ae6039eb","fe7279cc-a791-45be-80b6-3308032e909c"]
[debug] [XCUITest] Executing command 'getElementRect'
[debug] [W3C] Responding to client with driver.getElementRect() result: {"x":24,"y":162,"width":170,"height":53}
[HTTP] <-- GET /wd/hub/session/fe7279cc-a791-45be-80b6-3308032e909c/element/appium-image-element-bf442481-2c5d-4d36-9033-a358ae6039eb/rect 200 4 ms - 50
[HTTP] 
[HTTP] --> GET /wd/hub/session/fe7279cc-a791-45be-80b6-3308032e909c/element/appium-image-element-bf442481-2c5d-4d36-9033-a358ae6039eb/rect
[HTTP] {}
[debug] [W3C] Calling AppiumDriver.getElementRect() with args: ["appium-image-element-bf442481-2c5d-4d36-9033-a358ae6039eb","fe7279cc-a791-45be-80b6-3308032e909c"]
[debug] [XCUITest] Executing command 'getElementRect'
[debug] [W3C] Responding to client with driver.getElementRect() result: {"x":24,"y":162,"width":170,"height":53}
[HTTP] <-- GET /wd/hub/session/fe7279cc-a791-45be-80b6-3308032e909c/element/appium-image-element-bf442481-2c5d-4d36-9033-a358ae6039eb/rect 200 8 ms - 50
[HTTP] 
[HTTP] --> GET /wd/hub/session/fe7279cc-a791-45be-80b6-3308032e909c/element/appium-image-element-bf442481-2c5d-4d36-9033-a358ae6039eb/rect
[HTTP] {}
[debug] [W3C] Calling AppiumDriver.getElementRect() with args: ["appium-image-element-bf442481-2c5d-4d36-9033-a358ae6039eb","fe7279cc-a791-45be-80b6-3308032e909c"]
[debug] [XCUITest] Executing command 'getElementRect'
[debug] [W3C] Responding to client with driver.getElementRect() result: {"x":24,"y":162,"width":170,"height":53}
[HTTP] <-- GET /wd/hub/session/fe7279cc-a791-45be-80b6-3308032e909c/element/appium-image-element-bf442481-2c5d-4d36-9033-a358ae6039eb/rect 200 6 ms - 50
[HTTP] 
```

### CODE REVIEW

File: lib/basedriver/image-element.js [hunk 1]
```diff
@@ -14,6 +14,15 @@ const IMAGE_TAP_STRATEGIES = [
   IMAGE_EL_TAP_STRATEGY_W3C
 ];
 
+const AVAILABLE_COMMANDS = {
```
Comment (mykola-mokhnach) [inline lib/basedriver/image-element.js]: this can be an array as soon as all keys are equal to values

File: lib/basedriver/driver.js [hunk 2]
```diff
@@ -259,7 +259,7 @@ class BaseDriver extends Protocol {
     this.clearNewCommandTimeout();
 
     // if we don't have this command, it must not be implemented
-    if (!this[cmd]) {
+    if (!this[cmd] && !isImageCommand(args, cmd)) {
```
Comment (mykola-mokhnach) [inline lib/basedriver/driver.js]: we could use `getImgElFromArgs` instead

Comment (KazuCocoa) [reply lib/basedriver/driver.js]: ah, right.
`execute` in image-element.js raise `new errors.NotYetImplementedError();` if current command doesn't exist there.
e.g.: `element.tag_name` issues `/wd/hub/session/7446e842-fca2-4e11-a9ad-6e64cbe41e93/element/appium-image-element-48bd1a97-07b4-43c0-bc3e-5e044069c763/name` and calles `getName`. But it should raise `NotYetImplementedError`. We can get the raise in https://github.com/appium/appium-base-driver/pull/245/files#diff-69c68e3397852a86397892936887db6aR292

I see..

File: lib/basedriver/driver.js [hunk 3]
```diff
@@ -259,7 +259,7 @@ class BaseDriver extends Protocol {
     this.clearNewCommandTimeout();
 
     // if we don't have this command, it must not be implemented
-    if (!this[cmd]) {
+    if (!this[cmd] && !getImgElFromArgs(args)) {
```
Comment (mykola-mokhnach) [inline lib/basedriver/driver.js]: you could probably set the result of `getImgElFromArgs(args)` to a separate const variable, since it is called twice

**Comment by KazuCocoa** [comment]
cc @jlipps 
Let me merge this since currently `-image` doesn't work against some commands.
If you have something comments, feel free to comment/revert/change, etc

**Pull request merged**